linera-service/src/linera-exporter/storage.rs

```diff
@@ -0,0 +1,428 @@
+            / (size_of::<CryptoHash>() + size_of::<LiteBlockId>()) as u64;
+        let chain_states_cache = LfuCache::builder()
+            .max_capacity(chain_states_cache_capacity)
+            .build();
+
+        let mut shared_storage =
+            SharedStorage::new(storage, canonical_state, destination_states, limits);
+        let exporter_storage = ExporterStorage::new(shared_storage.clone()?);
+
+        Ok((
+            Self {
+                shared_storage,
+                chain_states_cache,
+                exporter_state_view: view,
+            },
+            exporter_storage,
+        ))
+    }
+
+    pub(super) async fn get_block(
+        &self,
+        hash: CryptoHash,
+    ) -> Result<Arc<GenericCertificate<ConfirmedBlock>>, ExporterError> {
+        self.shared_storage.get_block(hash).await
+    }
+
+    pub(super) async fn is_block_indexed(
+        &mut self,
+        block_id: &BlockId,
+    ) -> Result<bool, ExporterError> {
+        if let Some(status) = self.chain_states_cache.get(&block_id.chain_id) {
+            return Ok(status.height >= block_id.height);
+        }
+
+        if let Some(status) = self
+            .exporter_state_view
+            .get_chain_status(&block_id.chain_id)
+            .await?
+        {
+            let result = status.height >= block_id.height;
+            self.chain_states_cache.insert(block_id.chain_id, status);
+            return Ok(result);
+        }
+
+        Err(ExporterError::UnprocessedChain)
+    }
+
+    pub(super) async fn index_chain(&mut self, block_id: &BlockId) -> Result<(), ExporterError> {
+        ensure!(
+            block_id.height == BlockHeight::ZERO,
+            ExporterError::BadInitialization
+        );
+        self.shared_storage.push_block(block_id.hash).await?;
+        self.exporter_state_view
+            .initialize_chain(block_id.clone())
+            .await?;
+        self.chain_states_cache
+            .insert(block_id.chain_id, block_id.clone().into());
+
+        Ok(())
+    }
+
+    pub(super) async fn index_block(&mut self, block_id: &BlockId) -> Result<(), ExporterError> {
+        if block_id.height == BlockHeight::ZERO {
+            self.index_chain(block_id).await?;
+            return Ok(());
+        }
+
+        if self
+            .exporter_state_view
+            .index_block(block_id.clone())
+            .await?
+        {
+            self.shared_storage.push_block(block_id.hash).await?;
+            self.chain_states_cache
+                .insert(block_id.chain_id, block_id.clone().into());
+        }
+
+        Ok(())
+    }
+
+    pub(super) async fn save(&mut self) -> Result<(), ExporterError> {
+        let mut batch = Batch::new();
+        self.shared_storage
+            .shared_canonical_state
+            .flush(&mut batch)?;
+        self.exporter_state_view
+            .set_destination_states(self.shared_storage.destination_states.clone());
```

Comment:
It is a little bit odd to have the view used in this way.
Would it be possible to implement according to the same usage of `flush / save / rollback` as in other parts of the code?